clc/modules/object-storage/src/main/java/com/eucalyptus/objectstorage/ObjectStorageGateway.java

```diff
@@ -2013,113 +2055,317 @@ public DeleteBucketTaggingResponseType deleteBucketTagging(DeleteBucketTaggingTy
+
+      // Return all the CORS rule's allowed methods in the response
+      responseFields.setMethods(corsRuleMatch.getAllowedMethods());
+      
+      // Return all of the request's Access-Control-Request-Headers ni the response
+      responseFields.setAllowedHeaders(preflightRequest.getRequestHeaders());
+
+      // Return all of the CORS rule's expose headers in the response
+      responseFields.setExposeHeaders(corsRuleMatch.getExposeHeaders());
+      
+      responseFields.setMaxAgeSeconds(corsRuleMatch.getMaxAgeSeconds());
+
+      response.setStatus(HttpResponseStatus.OK);
+
+    } catch (S3Exception s3e) {
+      LOG.warn("Caught S3Exception while processing the preflight CORS request for bucket <" + 
+          bucketName + ">, CorrelationId: " + Contexts.lookup().getCorrelationId() + 
+          ", responding to client with: ", s3e);
+      throw s3e;
+    } catch (Exception ex) {
+      LOG.warn("Caught general exception while processing the preflight CORS request for bucket <" + 
+          bucketName + ">, CorrelationId: " + Contexts.lookup().getCorrelationId() + 
+          ", responding to client with 500 InternalError because of: ", ex);
+      throw new InternalErrorException(bucketName, ex);
+    }
+    
+    return response;
+  }
+
+  public static void addCorsResponseHeaders (MappingHttpResponse httpResponse) throws S3Exception {
+    LOG.debug("LPT In addCorsResponseHeaders (MappingHttpResponse)");
+
+    BaseMessage msg = (BaseMessage) httpResponse.getMessage();
+    addCorsResponseHeaders((HttpResponse) httpResponse, (BaseMessage) msg);
+  }
+  
+  public static void addCorsResponseHeaders (HttpResponse httpResponse, BaseMessage msg) throws S3Exception {
+    LOG.debug("LPT In addCorsResponseHeaders (HttpResponse, BaseMessage), msg class is " + msg.getClass());
+    if (msg != null && msg instanceof ObjectStorageResponseType) {
+      ObjectStorageResponseType response = (ObjectStorageResponseType) msg;
+      addCorsResponseHeaders(httpResponse, response.getOrigin(), response.getHttpMethod(), 
+          response.getBucket(), response.getBucketUuid());
+    } else if (msg instanceof ObjectStorageDataResponseType) {
+      ObjectStorageDataResponseType response = (ObjectStorageDataResponseType) msg;
+      addCorsResponseHeaders(httpResponse, response.getOrigin(), response.getHttpMethod(), 
+          response.getBucket(), response.getBucketUuid());
+    }
+  }
+  
+  public static void addCorsResponseHeaders (HttpResponse httpResponse, String origin, String httpMethod, String bucket, String bucketUuid) throws S3Exception {
+    LOG.debug("LPT In addCorsResponseHeaders (HttpResponse, origin, bucket, bucketUuid)");
+
+    LOG.debug("LPT Origin is: " + origin + ", Bucket (name or UUID) is: " + bucket + ", Bucket UUID is: " + bucketUuid);
+
+    if (origin != null && bucketUuid != null) {
+      List<CorsRule> corsRules;
+      try {
+        corsRules = BucketCorsManagers.getInstance().getCorsRules(bucketUuid);
```

Comment:
Will do. 
I was simplifying by populating the response headers directly in the HTTP fields in the outbound handlers, skipping the need to store them in new members of the response type classes and then copy them to the HTTP headers. But we discussed that doing the database lookups in the inbound handlers with the rest of the CORS database processing makes more sense and we might be able to reduce the amount of DB lookups that way.